### DISCUSSION TITLE
Generate Ninja project for an external dependency

### DISCUSSION THREAD

**Discussion by jeyblu** [microsoft/onnxruntime#10054] [General]
onnxruntime generates Visual Studio C++ projects and builds the binaries using Visual Studio C++ through cmake. We tried to get cmake to generate Ninja project to an onnxruntime external dependency, i.e. oneDNN. We want to try to use DPC++ for oneDNN which requires Ninja as described in the [documentation](https://oneapi-src.github.io/oneDNN/dev_guide_build.html#id1). We modified [dnnl.cmake ](https://github.com/microsoft/onnxruntime/blob/master/cmake/external/dnnl.cmake)to generate Ninja projects.

`set(DNNL_CPU_CMAKE_ARGS "-G Ninja " "-DDNNL_CPU_RUNTIME=DPCPP")`

We then tried to build onnxruntime.

`build.bat --config Debug --build_shared_lib --build_wheel --parallel --build_dir "build_dnnl_dpcpp" --use_dnnl --dnnl_cpu_runtime dpcpp --dnnl_gpu_runtime dpcpp --cmake_generator "Visual Studio 16 2019"`

In the middle of the build, an error was encountered.

> CUSTOMBUILD : CMake error : Could not create named generator Ninja [C:\code\onnxruntime\build_dnnl_dpcpp\Debug\project_dnnl.vcxproj]

> C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(241,5): error
 MSB8066: Custom build for 'C:\code\onnxruntime\build_dnnl_dpcpp\Debug\CMakeFiles\81be48453f9067b2e55640b018f10e08\project_dnnl-mk
dir.rule;C:\code\onnxruntime\build_dnnl_dpcpp\Debug\CMakeFiles\81be48453f9067b2e55640b018f10e08\project_dnnl-download.rule;C:\code
\onnxruntime\build_dnnl_dpcpp\Debug\CMakeFiles\81be48453f9067b2e55640b018f10e08\project_dnnl-update.rule;C:\code\onnxruntime\build
_dnnl_dpcpp\Debug\CMakeFiles\81be48453f9067b2e55640b018f10e08\project_dnnl-patch.rule;C:\code\onnxruntime\build_dnnl_dpcpp\Debug\C
MakeFiles\81be48453f9067b2e55640b018f10e08\project_dnnl-configure.rule;C:\code\onnxruntime\build_dnnl_dpcpp\Debug\CMakeFiles\81be4
8453f9067b2e55640b018f10e08\project_dnnl-build.rule;C:\code\onnxruntime\build_dnnl_dpcpp\Debug\CMakeFiles\81be48453f9067b2e55640b0
18f10e08\project_dnnl-install.rule;C:\code\onnxruntime\build_dnnl_dpcpp\Debug\CMakeFiles\c1660798fb8fb7a0bf6a3a07611cbc15\project_
dnnl-complete.rule;C:\code\onnxruntime\build_dnnl_dpcpp\Debug\CMakeFiles\eeeaecf442215b283257374763845981\project_dnnl.rule;C:\cod
e\onnxruntime\cmake\CMakeLists.txt' exited with code 1. [C:\code\onnxruntime\build_dnnl_dpcpp\Debug\project_dnnl.vcxproj]

**Comment by mszhanyi** [comment]
I agree. Visual C++ project building speed is slower than ninja.

**Comment by snnn** [comment]
Is it a cmake issue or onnx runtime issue?
